lib/wasi/src/runtime/resolver/inputs.rs

```diff
@@ -0,0 +1,339 @@
+        // the locator field
+        let (full_name, version) = match s.split_once('@') {
+            Some((n, v)) => (n, v),
+            None => (s, "*"),
+        };
+
+        let invalid_character = full_name
+            .char_indices()
+            .find(|(_, c)| !matches!(c, 'a'..='z' | 'A'..='Z' | '0'..='9' | '.'| '-'|'_' | '/'));
+        if let Some((index, c)) = invalid_character {
+            anyhow::bail!("Invalid character, {c:?}, at offset {index}");
+        }
+
+        let version = version
+            .parse()
+            .with_context(|| format!("Invalid version number, \"{version}\""))?;
+
+        Ok(PackageSpecifier::Registry {
+            full_name: full_name.to_string(),
+            version,
+        })
+    }
+}
+
+impl Display for PackageSpecifier {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            PackageSpecifier::Registry { full_name, version } => write!(f, "{full_name}@{version}"),
+            PackageSpecifier::Url(url) => Display::fmt(url, f),
+            PackageSpecifier::Path(path) => write!(f, "{}", path.display()),
+        }
+    }
+}
+
+/// A dependency constraint.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Dependency {
+    pub alias: String,
+    pub pkg: PackageSpecifier,
+}
+
+impl Dependency {
+    pub fn package_name(&self) -> Option<&str> {
+        match &self.pkg {
+            PackageSpecifier::Registry { full_name, .. } => Some(full_name),
+            _ => None,
+        }
+    }
+
+    pub fn alias(&self) -> &str {
+        &self.alias
+    }
+
+    pub fn version(&self) -> Option<&VersionReq> {
+        match &self.pkg {
+            PackageSpecifier::Registry { version, .. } => Some(version),
+            _ => None,
+        }
+    }
+}
+
+/// Some metadata a [`Source`][source] can provide about a package without
+/// needing to download the entire `*.webc` file.
+///
+/// [source]: crate::runtime::resolver::Source
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Summary {
+    pub pkg: PackageInfo,
+    pub dist: DistributionInfo,
+}
+
+impl Summary {
+    pub fn package_id(&self) -> PackageId {
+        self.pkg.id()
+    }
+
+    pub fn from_webc_file(path: impl AsRef<Path>) -> Result<Summary, Error> {
+        let path = path.as_ref().canonicalize()?;
+        let container = Container::from_disk(&path)?;
+        let webc_sha256 = WebcHash::for_file(&path)?;
+        let url = Url::from_file_path(&path).map_err(|_| {
+            anyhow::anyhow!("Unable to turn \"{}\" into a file:// URL", path.display())
+        })?;
+
+        let pkg = PackageInfo::from_manifest(container.manifest())?;
+        let dist = DistributionInfo {
+            webc: url,
+            webc_sha256,
+        };
+
+        Ok(Summary { pkg, dist })
+    }
+}
+
+/// Information about a package's contents.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct PackageInfo {
```

Comment:
How this type differs from the one in the manifest? Do we need a new one instead of unifying this two types?

Reply:
The manifest contains a lot of information, most of which is untyped (i.e. `serde_cbor::Value`), so I've introduced a trivial `PackageInfo` type because that lets us use just the information we need. 

Using a separate POD type also avoids coupling the resolution code too tightly with `webc::metadata::Manifest`, which is useful because I'd like to change several things in the `Manifest` after the 4.0 release.